### PR TITLE
fix(options): allow underscore in names

### PR DESCRIPTION
**Pull request by stanimirovv**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently underscores are turned into dashes.

Issue Number: https://github.com/nestjs/nest-cli/issues/489

## What is the new behavior?
Underscores are supported from nest-cli.
The commands built are with underscores, however, this will not fix the issue entirely since it must also be changed here:
https://github.com/nestjs/schematics/blob/2c732fb2718d7b33adfe64415d315aceee497c44/src/lib/controller/controller.factory.ts#L49

If the PR is accepted I will also make the PR in the other repo.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

### CODE REVIEW

File: lib/schematics/schematic.option.ts [hunk 1]
```diff
@@ -31,4 +28,13 @@ export class SchematicOption {
         return `${content}${char}`;
       }, '');
   }
+
+  private dasherize(str: string) {
+    const STRING_DASHERIZE_REGEXP = /[\s]/g;
```
Comment (micalevisk) [inline lib/schematics/schematic.option.ts:33]:
```suggestion
    const STRING_DASHERIZE_REGEXP = /\s/g;
```

**Comment by micalevisk** [comment]
you can take this one to fix the `new` command as well :smile: Currently, `nest new my_app` creates the `my-app` directory

https://github.com/nestjs/nest-cli/blob/73ba62898e7964d412ef58c3640bbb4307f54afe/actions/new.action.ts#L77

https://github.com/nestjs/nest-cli/blob/73ba62898e7964d412ef58c3640bbb4307f54afe/lib/package-managers/abstract.package-manager.ts#L26

in this case, you'll need to extract the `dasherize` method. Maybe to a new file called `lib/utils/formatting.ts`

**Comment by micalevisk** [comment]
@stanimirovv if you don't want to fix that one it's fine. Let us know.

**Comment by stanimirovv** [comment]
@micalevisk agreed, I will ping you when I've pushed the changes.

**Comment by stanimirovv** [comment]
@micalevisk I've applied the changes.

The fix still doesn't work from e2e point of view since it also needs to be adjusted in the schematics repo as well: https://github.com/nestjs/schematics/blob/2c732fb2718d7b33adfe64415d315aceee497c44/src/lib/controller/controller.factory.ts#L49

Once we have wrapped this PR I will move on to it.

**Comment by micalevisk** [comment]
indeed, I just notice this. That wrong behavior on `nest new my-app` is due to this line: https://github.com/nestjs/schematics/blob/2c732fb2718d7b33adfe64415d315aceee497c44/src/lib/application/application.factory.ts#L21 as the `dasherize` utility from `@angular-devkit/core` will turn `my_app` into `my-app` (and this is [the expected behavior](https://github.com/angular/angular-cli/blob/c4d3dfa14f5d145b831857645f27c056c893526a/packages/angular_devkit/core/src/utils/strings.ts#L38))

So the name `dasherize` of the function you wrote isn't good but naming is hard :smiling_face_with_tear:

File: lib/utils/formatting.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export function dasherize(str: string) {
```
Comment (micalevisk) [inline lib/utils/formatting.ts:1]:
```suggestion
export function keepKebabOrSnakeCase(str: string) {
```

maybe this name describes better?

---

```js
// kebab
dasherize('a-b-c') === 'a-b-c',
// snake
dasherize('a_b_c') === 'a_b_c',

dasherize('a b c') === 'a-b-c',
dasherize('a-b c') === 'a-b-c',
dasherize('a_b c') === 'a_b-c',
dasherize('a-b_c') === 'a-b_c',
```

Comment (stanimirovv) [reply lib/utils/formatting.ts:1]: I like the name, but the function transforms, right , I feel this needs to be showed in the name.

if you have camelCase it is transformed... hm... maybe something generic like transformCamelCase ? or transformName ? 

Honestly I don't really like my suggestions, but we deffinetly need to adjust the function name.

Comment (micalevisk) [reply lib/utils/formatting.ts:1]: I guess you could choose any other than just `dasherize`, and write JSDoc comment to describe its for now

```js
transformToKebabOrSnakeCase('fooBar') === 'foo-bar' // camel
transformToKebabOrSnakeCase('FooBar') === 'foo-bar' // pascal
transformToKebabOrSnakeCase('foo bar') === 'foo-bar' // spaced

// or maybe 'normalizeToKebabOrSnakeCase'
```

Comment (stanimirovv) [reply lib/utils/formatting.ts]: I really liked including 'normalize' in the name, nice suggestion.

I've updated the code. I am not really confident in the jsdoc description, but we can always adjust it.

File: lib/utils/formatting.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/**
+ *
+ * @param str
+ * @returns formated string
+ * @description normalizes input to supported path and file name format.
+ * Changes camelCase strings to kebab-case, replaces spaces with dash and keeps underscores.
+ */
+
```
Comment (micalevisk) [inline lib/utils/formatting.ts:8]:
```suggestion
```

just a minor formatting fix

File: lib/package-managers/abstract.package-manager.ts [hunk 4]
```diff
@@ -23,7 +23,7 @@ export abstract class AbstractPackageManager {
     try {
       const commandArgs = `${this.cli.install} ${this.cli.silentFlag}`;
       const collect = true;
-      const dasherizedDirectory = dasherize(directory);
+      const dasherizedDirectory = normalizeToKebabOrSnakeCase(directory);
```
Comment (micalevisk) [inline lib/package-managers/abstract.package-manager.ts:26]:
```suggestion
      const normalizedDirectory = normalizeToKebabOrSnakeCase(directory);
```

(and then rename the occurrences below)

Comment (stanimirovv) [reply lib/package-managers/abstract.package-manager.ts]: Nice find, I checked the other places - this was the only variable rename required.
I've applied a fix and pushed it.

**Review by micalevisk — APPROVED** [review]

**Comment by stanimirovv** [comment]
@micalevisk no action is required on my part as of now, right ?

**Comment by micalevisk** [comment]
@stanimirovv right :) thank you. You can start the fix on https://github.com/nestjs/schematics side

**Comment by kamilmysliwiec** [comment]
LGTM

**Pull request merged**